### ISSUE TITLE
Adaptation support / example for multiple chains

### ISSUE THREAD

**Issue by martiningram** [blackjax-devs/blackjax#156]
Dear blackjax team,

Thank you for your hard work, this project is really cool!

I've been playing around with the PyMC support, and I think I've managed to update [the example notebook](https://github.com/blackjax-devs/blackjax/blob/main/notebooks/use_with_pymc3.ipynb) to use PyMC v4 and `aesara`. This wasn't very hard to do thankfully, and I'll be happy to contribute it as a new example / update the old one if that's of interest.

However, my ultimate aim would be to include blackjax as part of the recent MCMC speed comparisons (see [code](https://github.com/martiningram/mcmc_runtime_comparison) and [blog post](https://www.pymc-labs.io/blog-posts/pymc-stan-benchmark/)). To get the best out of blackjax, I'd like to run four chains in parallel like with the other methods. That's where I've run into a bit of trouble! There's an example of _sampling_ multiple chains in the Intro notebook [here](https://github.com/blackjax-devs/blackjax/blob/main/notebooks/Introduction.ipynb) but I don't see anything about also doing separate _adaptation_...! I'm actually not sure what the best way to go is there -- should chains be adapted separately or one step size and mass matrix should be estimated and shared in a warmup (probably)?

In any case, I feel like an example of how best to do adaption with multiple chains would be helpful to add. Please let me know if I missed an existing example somewhere, or if there is already a recommended approach.

Thanks for your help,
Martin

**Comment by junpenglao** [comment]
Hi Martin,
You can take a look at the test here: https://github.com/blackjax-devs/blackjax/blob/main/tests/test_sampling.py#L66
`self.variant` is the place you can put `jax.vmap` or `jax.pmap`. In the test the warmup and sampling is ran separately but you can put everything in a single function 

Using either vmap or pmap chains will be adapted separately - vmap is in general faster but if you have heavy tail posterior (a chain might need a lot of steps with some probability) pmap will likely be faster

**Comment by rlouf** [comment]
@junpenglao said it all. Also note there's work in progress to improve the warmup performance.

**Comment by martiningram** [comment]
Thank you, @junpenglao and @rlouf for your quick response!

Your suggestion does indeed work. I was able to take inspiration from the code snippet to modify my own code. Here's what I've come up with.

First, get the initial state and the `logp_fn`:

```python
chains = 4
n_warmup = 1000
n_samples = 50_000

seed = jax.random.PRNGKey(1234)

# Interface with PyMC v4
logp_fn_jax = get_jaxified_logp(model)

rv_names = [rv.name for rv in model.value_vars]
initial_point = model.recompute_initial_point()
init_state = [initial_point[rv_name] for rv_name in rv_names]
init_state_batched = jax.tree_map(lambda x: np.repeat(x[None, ...], chains, axis=0), init_state)

initial_state = jax.vmap(nuts.new_state, in_axes=(0, None))(init_state_batched, logp_fn_jax)
```

Do the warmup via `vmap`:

```python
# Warmup
kernel_factory = lambda step_size, inverse_mass_matrix: nuts.kernel(
    logp_fn_jax, step_size, inverse_mass_matrix
)

init_seeds = jax.random.split(seed, 4)

warmup_fn = lambda seed, init_state: stan_warmup.run(seed, kernel_factory, init_state, n_warmup)

last_state, (step_size, inverse_mass_matrix), _ = jax.vmap(warmup_fn)(init_seeds, initial_state)
```

Finally, sample, also with `vmap` (is this the best way?):

```python
# Sampling
def inner_inference_loop(
    rng_key, step_size, inverse_mass_matrix, initial_state, num_samples, kernel_factory
):
    
    kernel = kernel_factory(step_size, inverse_mass_matrix)
    
    def one_step(state, rng_key):
        
        key = jax.random.split(rng_key, 2)
        
        state, info = kernel(key, state)
        
        return state, (state, info)
    
    keys = jax.random.split(rng_key, num_samples)
    
    _, states = jax.lax.scan(one_step, initial_state, keys)
    
    return states

from functools import partial

to_vmap = partial(inner_inference_loop, num_samples=n_samples, kernel_factory=kernel_factory)

# TODO: need a new seed maybe?
start_keys = jax.random.split(seed, 4)

results = jax.vmap(to_vmap)(start_keys, step_size, inverse_mass_matrix, last_state)
```

The notebook with this code is [available here](https://github.com/martiningram/blackjax/blob/main/notebooks/Use%20with%20PyMC%20v4.ipynb). Let me know what you think and what I should fix!

Cheers,
Martin

**Comment by martiningram** [comment]
Just a quick update: something I'm doing is probably wrong, because I'm [getting weird results on eight schools](https://github.com/martiningram/blackjax/blob/main/notebooks/Use%20with%20PyMC%20v4.ipynb). It looks like maybe the inverse mass matrix is weird; I'm getting:

```
ass matrix from chain 1
[9.90173580e-06 9.90233208e-06 3.22715215e-03 9.90126116e-06
 9.90164561e-06 9.90344630e-06 9.90159643e-06 9.90119435e-06
 9.90264960e-06 2.66199793e+04]
Mass matrix from chain 2
[9.90424376e-06 9.90265363e-06 9.90209812e-06 9.90184892e-06
 9.90331472e-06 2.45332112e+04 9.90261137e-06 9.91300716e-06
 9.90238655e-06 9.90635319e-06]
Mass matrix from chain 3
[9.90131336e-06 9.90168204e-06 9.90201598e-06 9.90134089e-06
 9.90215500e-06 1.15046654e-05 9.90170771e-06 9.90129017e-06
 9.90137060e-06 1.78055500e+04]
Mass matrix from chain 4
[9.90143324e-06 9.90434228e-06 9.90339854e-06 2.27641479e+04
 9.90208592e-06 2.84121890e-05 9.90229951e-06 9.90304749e-06
 8.14807809e-02 9.90586300e-06]
```
So they are quite different from one chain to the next...

I'd be grateful for pointers if anyone can see what I'm doing wrong!

**Comment by junpenglao** [comment]
We suspect there is a bug somewhere (@rlouf has more detail?)

**Comment by rlouf** [comment]
Yes, I suspect the "average acceptance probability" returned by the NUTS sampler is not correct. I've been procrastinating on this for too long and will try to find time the following evenings or this weekend to look into it.

**Comment by rlouf** [comment]
Hey @martiningram lots of work since I last replied on #153 and #159. We've identified one major difference with Numpyro which likely explains the perf difference, will keep you posted!

**Comment by martiningram** [comment]
Awesome, thanks for the update @rlouf ! Looking forward to trying the new code when it's ready :)

**Comment by rlouf** [comment]
You're welcome! You'll hopefully find the user interface more pleasant as well :)

**Comment by rlouf** [comment]
Hey @martiningram is this still an issue with the latest version?

**Comment by martiningram** [comment]
Hi @rlouf , thanks for following up! I'm not sure, unfortunately. I tried running the old notebook I wrote (https://github.com/martiningram/blackjax/blob/main/notebooks/Use%20with%20PyMC%20v4.ipynb) but it looks like the API has changed a bit, so I'd have to make some updates. I'm not sure I'll have time to do that in the near future, unfortunately. But I think the issue could be closed if we have an example of multi-chain adaptation somewhere; is that a thing already?

Thanks and all the best,
Martin
